File: lead-validation-python/REST/lv_response.py
from dataclasses import dataclass
from typing import Optional, List


@dataclass
class InformationComponent:
    """Information Component for API responses."""
    Name: Optional[str] = None
    Value: Optional[str] = None

    def __str__(self) -> str:
        return f"Name={self.Name}, Value={self.Value}"


@dataclass
class PhoneContact:
    """Phone Contact for lead validation responses."""
    Name: Optional[str] = None
    Address: Optional[str] = None
    City: Optional[str] = None
    State: Optional[str] = None
    Zip: Optional[str] = None
    Type: Optional[str] = None

    def __str__(self) -> str:
        return (f"PhoneContact: Name={self.Name}, Address={self.Address}, City={self.City}, "
                f"State={self.State}, Zip={self.Zip}, Type={self.Type}")


@dataclass
class Error:
    """Error object for API responses."""
    Type: Optional[str] = None
    TypeCode: Optional[str] = None
    Desc: Optional[str] = None
    DescCode: Optional[str] = None

    def __str__(self) -> str:
        return (f"Error: Type={self.Type}, TypeCode={self.TypeCode}, "
                f"Desc={self.Desc}, DescCode={self.DescCode}")


@dataclass
class LVResponse:
    """Response from ValidateLead_V3 API, containing lead validation information."""
    OverallCertainty: Optional[str] = None
    OverallQuality: Optional[str] = None
    LeadType: Optional[str] = None
    LeadCountry: Optional[str] = None
    NoteCodes: Optional[str] = None
    NoteDesc: Optional[str] = None
    NameCertainty: Optional[str] = None
    NameQuality: Optional[str] = None
    FirstName: Optional[str] = None
    LastName: Optional[str] = None
    FirstNameClean: Optional[str] = None
    LastNameClean: Optional[str] = None
    NameNoteCodes: Optional[str] = None
    NameNoteDesc: Optional[str] = None
    AddressCertainty: Optional[str] = None
    AddressQuality: Optional[str] = None
    Address1: Optional[str] = None
    Address2: Optional[str] = None
    Address3: Optional[str] = None
    Address4: Optional[str] = None
    Address5: Optional[str] = None
    AddressLocality: Optional[str] = None
    AddressAdminArea: Optional[str] = None
    AddressPostalCode: Optional[str] = None
    AddressCountry: Optional[str] = None
    AddressNoteCodes: Optional[str] = None
    AddressNoteDesc: Optional[str] = None
    EmailCertainty: Optional[str] = None
    EmailQuality: Optional[str] = None
    EmailCorrected: Optional[str] = None
    EmailNoteCodes: Optional[str] = None
    EmailNoteDesc: Optional[str] = None
    IPAddressCertainty: Optional[str] = None
    IPAddressQuality: Optional[str] = None
    IPCountry: Optional[str] = None
    IPLocality: Optional[str] = None
    IPAdminArea: Optional[str] = None
    IPNoteCodes: Optional[str] = None
    IPNoteDesc: Optional[str] = None
    Phone1Certainty: Optional[str] = None
    Phone1Quality: Optional[str] = None
    Phone1Locality: Optional[str] = None
    Phone1AdminArea: Optional[str] = None
    Phone1Country: Optional[str] = None
    Phone1NoteCodes: Optional[str] = None
    Phone1NoteDesc: Optional[str] = None
    Phone2Certainty: Optional[str] = None
    Phone2Quality: Optional[str] = None
    Phone2Locality: Optional[str] = None
    Phone2AdminArea: Optional[str] = None
    Phone2Country: Optional[str] = None
    Phone2NoteCodes: Optional[str] = None
    Phone2NoteDesc: Optional[str] = None
    PhoneContact: Optional['PhoneContact'] = None
    InformationComponents: Optional[List['InformationComponent']] = None
    Error: Optional['Error'] = None

    def __post_init__(self):
        if self.InformationComponents is None:
            self.InformationComponents = []

    def __str__(self) -> str:
        info_components_string = ', '.join(str(component) for component in self.InformationComponents) if self.InformationComponents else 'None'
        phone_contact = str(self.PhoneContact) if self.PhoneContact else 'None'
        error = str(self.Error) if self.Error else 'None'
        return (f"LVResponse: OverallCertainty={self.OverallCertainty}, OverallQuality={self.OverallQuality}, "
                f"LeadType={self.LeadType}, LeadCountry={self.LeadCountry}, NoteCodes={self.NoteCodes}, NoteDesc={self.NoteDesc}, "
                f"NameCertainty={self.NameCertainty}, NameQuality={self.NameQuality}, FirstName={self.FirstName}, LastName={self.LastName}, "
                f"FirstNameClean={self.FirstNameClean}, LastNameClean={self.LastNameClean}, NameNoteCodes={self.NameNoteCodes}, NameNoteDesc={self.NameNoteDesc}, "
                f"AddressCertainty={self.AddressCertainty}, AddressQuality={self.AddressQuality}, Address1={self.Address1}, Address2={self.Address2}, "
                f"Address3={self.Address3}, Address4={self.Address4}, Address5={self.Address5}, AddressLocality={self.AddressLocality}, "
                f"AddressAdminArea={self.AddressAdminArea}, AddressPostalCode={self.AddressPostalCode}, AddressCountry={self.AddressCountry}, "
                f"AddressNoteCodes={self.AddressNoteCodes}, AddressNoteDesc={self.AddressNoteDesc}, EmailCertainty={self.EmailCertainty}, "
                f"EmailQuality={self.EmailQuality}, EmailCorrected={self.EmailCorrected}, EmailNoteCodes={self.EmailNoteCodes}, EmailNoteDesc={self.EmailNoteDesc}, "
                f"IPAddressCertainty={self.IPAddressCertainty}, IPAddressQuality={self.IPAddressQuality}, IPCountry={self.IPCountry}, "
                f"IPLocality={self.IPLocality}, IPAdminArea={self.IPAdminArea}, IPNoteCodes={self.IPNoteCodes}, IPNoteDesc={self.IPNoteDesc}, "
                f"Phone1Certainty={self.Phone1Certainty}, Phone1Quality={self.Phone1Quality}, Phone1Locality={self.Phone1Locality}, "
                f"Phone1AdminArea={self.Phone1AdminArea}, Phone1Country={self.Phone1Country}, Phone1NoteCodes={self.Phone1NoteCodes}, Phone1NoteDesc={self.Phone1NoteDesc}, "
                f"Phone2Certainty={self.Phone2Certainty}, Phone2Quality={self.Phone2Quality}, Phone2Locality={self.Phone2Locality}, "
                f"Phone2AdminArea={self.Phone2AdminArea}, Phone2Country={self.Phone2Country}, Phone2NoteCodes={self.Phone2NoteCodes}, Phone2NoteDesc={self.Phone2NoteDesc}, "
                f"PhoneContact={phone_contact}, InformationComponents=[{info_components_string}], Error={error}")

File: lead-validation-nodejs/REST/lv_response.js
/**
 * Information Component for API responses.
 */
export class InformationComponent {
    constructor(data = {}) {
        this.Name = data.Name || null;
        this.Value = data.Value || null;
    }

    toString() {
        return `Name = ${this.Name}, Value = ${this.Value}`;
    }
}

/**
 * Phone Contact for lead validation responses.
 */
export class PhoneContact {
    constructor(data = {}) {
        this.Name = data.Name || null;
        this.Address = data.Address || null;
        this.City = data.City || null;
        this.State = data.State || null;
        this.Zip = data.Zip || null;
        this.Type = data.Type || null;
    }

    toString() {
        return `PhoneContact: Name = ${this.Name}, Address = ${this.Address}, City = ${this.City}, State = ${this.State}, Zip = ${this.Zip}, Type = ${this.Type}`;
    }
}

/**
 * Error object for API responses.
 */
export class Error {
    constructor(data = {}) {
        this.Type = data.Type || null;
        this.TypeCode = data.TypeCode || null;
        this.Desc = data.Desc || null;
        this.DescCode = data.DescCode || null;
    }

    toString() {
        return `Error: Type = ${this.Type}, TypeCode = ${this.TypeCode}, Desc = ${this.Desc}, DescCode = ${this.DescCode}`;
    }
}

/**
 * Response from ValidateLead_V3 API, containing lead validation information.
 */
export class LVResponse {
    constructor(data = {}) {
        this.OverallCertainty = data.OverallCertainty || null;
        this.OverallQuality = data.OverallQuality || null;
        this.LeadType = data.LeadType || null;
        this.LeadCountry = data.LeadCountry || null;
        this.NoteCodes = data.NoteCodes || null;
        this.NoteDesc = data.NoteDesc || null;
        this.NameCertainty = data.NameCertainty || null;
        this.NameQuality = data.NameQuality || null;
        this.FirstName = data.FirstName || null;
        this.LastName = data.LastName || null;
        this.FirstNameClean = data.FirstNameClean || null;
        this.LastNameClean = data.LastNameClean || null;
        this.NameNoteCodes = data.NameNoteCodes || null;
        this.NameNoteDesc = data.NameNoteDesc || null;
        this.AddressCertainty = data.AddressCertainty || null;
        this.AddressQuality = data.AddressQuality || null;
        this.Address1 = data.Address1 || null;
        this.Address2 = data.Address2 || null;
        this.Address3 = data.Address3 || null;
        this.Address4 = data.Address4 || null;
        this.Address5 = data.Address5 || null;
        this.AddressLocality = data.AddressLocality || null;
        this.AddressAdminArea = data.AddressAdminArea || null;
        this.AddressPostalCode = data.AddressPostalCode || null;
        this.AddressCountry = data.AddressCountry || null;
        this.AddressNoteCodes = data.AddressNoteCodes || null;
        this.AddressNoteDesc = data.AddressNoteDesc || null;
        this.EmailCertainty = data.EmailCertainty || null;
        this.EmailQuality = data.EmailQuality || null;
        this.EmailCorrected = data.EmailCorrected || null;
        this.EmailNoteCodes = data.EmailNoteCodes || null;
        this.EmailNoteDesc = data.EmailNoteDesc || null;
        this.IPAddressCertainty = data.IPAddressCertainty || null;
        this.IPAddressQuality = data.IPAddressQuality || null;
        this.IPCountry = data.IPCountry || null;
        this.IPLocality = data.IPLocality || null;
        this.IPAdminArea = data.IPAdminArea || null;
        this.IPNoteCodes = data.IPNoteCodes || null;
        this.IPNoteDesc = data.IPNoteDesc || null;
        this.Phone1Certainty = data.Phone1Certainty || null;
        this.Phone1Quality = data.Phone1Quality || null;
        this.Phone1Locality = data.Phone1Locality || null;
        this.Phone1AdminArea = data.Phone1AdminArea || null;
        this.Phone1Country = data.Phone1Country || null;
        this.Phone1NoteCodes = data.Phone1NoteCodes || null;
        this.Phone1NoteDesc = data.Phone1NoteDesc || null;
        this.Phone2Certainty = data.Phone2Certainty || null;
        this.Phone2Quality = data.Phone2Quality || null;
        this.Phone2Locality = data.Phone2Locality || null;
        this.Phone2AdminArea = data.Phone2AdminArea || null;
        this.Phone2Country = data.Phone2Country || null;
        this.Phone2NoteCodes = data.Phone2NoteCodes || null;
        this.Phone2NoteDesc = data.Phone2NoteDesc || null;
        this.PhoneContact = data.PhoneContact ? new PhoneContact(data.PhoneContact) : null;
        this.InformationComponents = (data.InformationComponents || []).map(component => new InformationComponent(component));
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        const infoComponentsString = this.InformationComponents.length
            ? this.InformationComponents.map(component => component.toString()).join(', ')
            : 'null';
        return `LVResponse: OverallCertainty = ${this.OverallCertainty}, OverallQuality = ${this.OverallQuality}, ` +
            `LeadType = ${this.LeadType}, LeadCountry = ${this.LeadCountry}, NoteCodes = ${this.NoteCodes}, NoteDesc = ${this.NoteDesc}, ` +
            `NameCertainty = ${this.NameCertainty}, NameQuality = ${this.NameQuality}, FirstName = ${this.FirstName}, LastName = ${this.LastName}, ` +
            `FirstNameClean = ${this.FirstNameClean}, LastNameClean = ${this.LastNameClean}, NameNoteCodes = ${this.NameNoteCodes}, NameNoteDesc = ${this.NameNoteDesc}, ` +
            `AddressCertainty = ${this.AddressCertainty}, AddressQuality = ${this.AddressQuality}, Address1 = ${this.Address1}, Address2 = ${this.Address2}, ` +
            `Address3 = ${this.Address3}, Address4 = ${this.Address4}, Address5 = ${this.Address5}, AddressLocality = ${this.AddressLocality}, ` +
            `AddressAdminArea = ${this.AddressAdminArea}, AddressPostalCode = ${this.AddressPostalCode}, AddressCountry = ${this.AddressCountry}, ` +
            `AddressNoteCodes = ${this.AddressNoteCodes}, AddressNoteDesc = ${this.AddressNoteDesc}, EmailCertainty = ${this.EmailCertainty}, ` +
            `EmailQuality = ${this.EmailQuality}, EmailCorrected = ${this.EmailCorrected}, EmailNoteCodes = ${this.EmailNoteCodes}, EmailNoteDesc = ${this.EmailNoteDesc}, ` +
            `IPAddressCertainty = ${this.IPAddressCertainty}, IPAddressQuality = ${this.IPAddressQuality}, IPCountry = ${this.IPCountry}, ` +
            `IPLocality = ${this.IPLocality}, IPAdminArea = ${this.IPAdminArea}, IPNoteCodes = ${this.IPNoteCodes}, IPNoteDesc = ${this.IPNoteDesc}, ` +
            `Phone1Certainty = ${this.Phone1Certainty}, Phone1Quality = ${this.Phone1Quality}, Phone1Locality = ${this.Phone1Locality}, ` +
            `Phone1AdminArea = ${this.Phone1AdminArea}, Phone1Country = ${this.Phone1Country}, Phone1NoteCodes = ${this.Phone1NoteCodes}, Phone1NoteDesc = ${this.Phone1NoteDesc}, ` +
            `Phone2Certainty = ${this.Phone2Certainty}, Phone2Quality = ${this.Phone2Quality}, Phone2Locality = ${this.Phone2Locality}, ` +
            `Phone2AdminArea = ${this.Phone2AdminArea}, Phone2Country = ${this.Phone2Country}, Phone2NoteCodes = ${this.Phone2NoteCodes}, Phone2NoteDesc = ${this.Phone2NoteDesc}, ` +
            `PhoneContact = ${this.PhoneContact ? this.PhoneContact.toString() : 'null'}, InformationComponents = [${infoComponentsString}], ` +
            `Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}

export default LVResponse;